### PR TITLE
Benchmarks multi stage

### LINKED ISSUE #197: add benchmark tests for multistage
- dynamic - static
- dynamic - dynamic

- 3d embankment train initialisation and movement
- changing material model between stages

### PR DESCRIPTION
**Pull request by morettid**
Benchmarks are added for:

- Multi-stage with switching the material for a UMAT and halving the stiffness under sinus load
- Multi-stage with a pulse and switching from dynamic to static after the oscillations are dissipated

Resolves #197

### CODE REVIEW

File: benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    # create new stage
+    model_stage_2 = stem.create_new_stage(0.01, 0.1)
+
+    # Set up solver settings for the new stage
+    model_stage_2.project_parameters.settings.solution_type = SolutionType.QUASI_STATIC
+    model_stage_2.project_parameters.settings.rayleigh_k = 0.0
+    model_stage_2.project_parameters.settings.rayleigh_m = 0.0
+
+    # model_stage_2.process_model_parts[0].parameters.value = [0, -1050, 0]
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py:132]: I think its better to keep this, as you can see if something changes, else dynamic and quasi static hold the same results

File: benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    # Set up problem data
+    problem = Problem(problem_name="test_multi_stage_dynamic_to_static", number_of_threads=2, settings=solver_settings)
+    model_stage_1.project_parameters = problem
+
+    # Define the results to be written to the output file
+    # Nodal results
+    nodal_results = [NodalOutput.DISPLACEMENT, NodalOutput.VELOCITY]
+
+    # Define the output process
+    model_stage_1.add_output_settings(output_parameters=VtkOutputParameters(file_format="ascii",
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py:112]: can you reduce the amount of output files, or check a node from a json file

File: benchmark_tests/test_multi_stage_umat/test_multi_stage_umat.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    # Set up problem data
+    problem = Problem(problem_name="test_multi_stage_umat", number_of_threads=2, settings=solver_settings)
+    model_stage_1.project_parameters = problem
+
+    # Define the results to be written to the output file
+    # Nodal results
+    nodal_results = [NodalOutput.DISPLACEMENT, NodalOutput.VELOCITY]
+
+    # Define the output process
+    model_stage_1.add_output_settings(output_parameters=VtkOutputParameters(file_format="ascii",
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_umat/test_multi_stage_umat.py:126]: can you reduce the amount of output files are check a node in a json?

File: benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    # create new stage
+    delta_time_stage_2 = 0.01
+    model_stage_2 = stem.create_new_stage(delta_time_stage_2, 0.03)
+
+    # Set up solver settings for the new stage
+    model_stage_2.project_parameters.settings.solution_type = SolutionType.QUASI_STATIC
+    model_stage_2.project_parameters.settings.rayleigh_k = 0.0
+    model_stage_2.project_parameters.settings.rayleigh_m = 0.0
+
+    # model_stage_2.process_model_parts[0].parameters.value = [0, -1050, 0]
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py:173]: uncomment

File: benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            label="Expected",
+        )
+        ax[0].plot(
+            merged_calculated_data["TIME"],
+            merged_calculated_data["NODE_5"]["DISPLACEMENT_X"],
+            label="Calculated",
+        )
+
+        ax[1].set_title("Displacements Y")
+        ax[0].set_ylabel("d_x [m]")
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py:228]:
```suggestion
        ax[0].set_ylabel("displacement_y [m]")
```

File: benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    merged_expected_data = Utils.merge(expected_data_stage1, expected_data_stage2)
+    merged_calculated_data = Utils.merge(calculated_data_stage1, calculated_data_stage2)
+
+    # Only calculate analytical solution and show results if SHOW_RESULTS is True
+    if SHOW_RESULTS:
+        import matplotlib.pyplot as plt
+
+        fig, ax = plt.subplots(2, 1, figsize=(8, 6), sharex="all")
+        ax[0].set_title("Displacements X")
+        ax[0].set_ylabel("d_x [m]")
```
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_dynamic_to_static/test_multistage_dynamic_to_static.py:215]:
```suggestion
        ax[0].set_ylabel("displacement_x [m]")
```

File: benchmark_tests/test_multi_stage_umat/linear_elastic.dll
Comment (aronnoordam) [inline benchmark_tests/test_multi_stage_umat/linear_elastic.dll:1]: move linear elastic dll to a shared location, such that we dont have multiple copies in the repo

**Comment by morettid** [comment]
image

Local vs Remote results differ. Maybe something related to version of gmsh or something like that?

**Comment by morettid** [comment]
@aronnoordam  should be okay now!

**Review by aronnoordam — APPROVED** [review]

**Pull request merged**